File: agent_service/blackboard.py
"""
Blackboard — 所有 Agent 共享的读写黑板。
Agent 透过黑板看到彼此的输出，形成真正的协作而非独立管道。
"""


class Blackboard:
    def __init__(self):
        self._sections = {
            "current": {},       # 当前笔记元信息
            "vault": {},         # vault 全局信息
            "similar": [],       # 语义相近笔记
            "findings": {        # 各 Agent 的分析产出
                "links": [],     # LinkWeaver 产出
                "concepts": [],  # LinkWeaver 产出
                "orphans": [],   # LinkWeaver 产出
                "tags": [],      # StructureGuardian 产出
                "structure": {}, # StructureGuardian 产出
            },
            "review": {          # Reviewer 最终产出
                "suggestions": [],
                "summary": "",
            },
            "session": {         # 会话级信息
                "user_preferences": {},
                "rejected_this_session": [],
                "hook_points": {},  # 留给后续 Agent 之间插入消息
            },
        }

    def read(self, section: str) -> dict:
        return self._sections.get(section, {})

    def write(self, section: str, data):
        self._sections[section] = data

    def update(self, section: str, data: dict):
        target = self._sections.get(section, {})
        if isinstance(target, dict):
            target.update(data)
        elif isinstance(target, list):
            if isinstance(data, list):
                target.extend(data)
            else:
                target.append(data)

    def write_finding(self, agent: str, key: str, data):
        self._sections["findings"][key] = data

    def snapshot(self) -> dict:
        return self._sections

    def has(self, section: str) -> bool:
        return bool(self._sections.get(section))

    def clear_session(self):
        self._sections["session"] = {
            "user_preferences": {},
            "rejected_this_session": [],
            "hook_points": {},
        }

File: obsidian-plugin/src/blackboard.ts
/**
 * Blackboard — 所有 Agent 共享的读写黑板。
 * Agent 透过黑板看到彼此的输出，形成真正的协作而非独立管道。
 */

export interface NoteContext {
  noteId: string;
  filePath: string;
  title: string;
  content: string;
  tags: string[];
}

export interface VaultInfo {
  totalNotes: number;
  allTitles: TitleEntry[];
  existingTags: string[];
}

export interface TitleEntry {
  title: string;
  path: string;
  dir: string;
  tags: string[];
}

export interface SimilarNote {
  title: string;
  path: string;
  distance: number;
  tags: string[];
}

export interface Findings {
  // 情报员产出
  contextSummary?: string;
  keyEntities?: string[];
  dataQuality?: string;
  scoutNotes?: string;
  // 链接师产出
  links?: LinkSuggestion[];
  concepts?: string[];
  orphans?: OrphanNote[];
  linkNotes?: string;
  // 架构师产出
  tags?: TagSuggestion[];
  structure?: StructureAnalysis;
  structureNotes?: string;
}

export interface LinkSuggestion {
  target: string;
  anchor_text: string;
  reason: string;
  confidence: number;
}

export interface OrphanNote {
  note_title: string;
  reason: string;
}

export interface TagSuggestion {
  current: string;
  suggested: string;
  reason: string;
}

export interface StructureAnalysis {
  frontmatter?: {
    is_present?: boolean;
    missing_fields?: string[];
    issues?: string[];
  };
  split?: {
    needs_split?: boolean;
    reason?: string;
    suggested_topics?: string[];
  };
  merge?: {
    needs_merge?: boolean;
    reason?: string;
    candidates?: string[];
  };
  moc?: {
    needs_moc?: boolean;
    topic?: string;
    reason?: string;
  };
}

export interface ReviewResult {
  suggestions: FinalSuggestion[];
  summary: string;
  conflicts?: ConflictNote[];
  notes?: string;
}

export interface FinalSuggestion {
  type: string;
  priority: number;
  title: string;
  description: string;
  confidence?: number;
  source_agent?: string;
  is_conflicted?: boolean;
  conflict_note?: string;
}

export interface ConflictNote {
  suggestion_a: string;
  suggestion_b: string;
  resolution: string;
}

interface SessionInfo {
  userPreferences: Record<string, unknown>;
  rejectedThisSession: string[];
  hookPoints: Record<string, unknown>;
}

interface BlackboardData {
  current: Partial<NoteContext>;
  vault: VaultInfo;
  similar: SimilarNote[];
  related: RelatedNoteProfile[];   // 语义库返回的相近笔记
  context: {
    sameDir: TitleEntry[];
    matched: TitleEntry[];
  };
  findings: Findings;
  review: ReviewResult;
  session: SessionInfo;
}

export interface RelatedNoteProfile {
  path: string;
  title: string;
  summary: string;
  keyTopics: string[];
  score: number;
  reason: string;
}

export class Blackboard {
  private data: BlackboardData = {
    current: {},
    vault: { totalNotes: 0, allTitles: [], existingTags: [] },
    similar: [],
    related: [],
    context: { sameDir: [], matched: [] },
    findings: {},
    review: { suggestions: [], summary: "" },
    session: {
      userPreferences: {},
      rejectedThisSession: [],
      hookPoints: {},
    },
  };

  read<K extends keyof BlackboardData>(section: K): BlackboardData[K] {
    return this.data[section];
  }

  write<K extends keyof BlackboardData>(section: K, value: BlackboardData[K]): void {
    this.data[section] = value;
  }

  updateFindings(partial: Partial<Findings>): void {
    Object.assign(this.data.findings, partial);
  }

  has<K extends keyof BlackboardData>(section: K): boolean {
    const val = this.data[section];
    if (Array.isArray(val)) return val.length > 0;
    if (typeof val === "object") return Object.keys(val).length > 0;
    return Boolean(val);
  }

  clearSession(): void {
    this.data.session = {
      userPreferences: {},
      rejectedThisSession: [],
      hookPoints: {},
    };
  }
}
